peps/pep-0800.rst

```diff
@@ -0,0 +1,352 @@
+stubs could condition the presence of the ``@solid_base`` decorator on the implementation where necessary.
+
+``@solid_base`` in implementation files
+---------------------------------------
+
+The most obvious use case for the ``@solid_base`` decorator will be in stub files for C libraries, such as the standard library,
+for marking solid bases implemented in C.
+
+However, there are also use cases for marking solid bases in implementation files, where the effect would be to disallow
+the existence of child classes that inherit from the decorated class and another solid base, such as a standard library class
+or another user class decorated with ``@solid_base``. For example, this could allow type checkers to flag code that can only
+be reachable if a class exists that inherits from both a user class and a standard library class such as ``int`` or ``str``,
+which may be technically possible but not practically plausible.
+
+::
+
+    @solid_base
+    class BaseModel:
+        # ... General logic for model classes
+        pass
+
+    class Species(BaseModel):
+        name: str
+        # ... more fields
+
+    def process_species(species: Species):
+        if isinstance(species, str):  # oops, forgot `.name`
+            pass  # type checker should warn about this branch being unreachable
+            # BaseModel and str are solid bases, so a class that inherits from both cannot exist
+
+This is similar in principle to the existing ``@final`` decorator, which also acts to restrict subclassing: in stubs, it
+is used to mark classes that programmatically disallow subclassing, but in implementation files, it is often used to
+indicate that a class is not intended to be subclassed, without runtime enforcement.
+
+``@solid_base`` on special classes
+----------------------------------
+
+The ``@solid_base`` decorator is primarily intended for nominal classes, but the type system contains some other constructs that
+syntactically use class definitions, so we have to consider whether the decorator should be allowed on them as well, and if so,
+what it means.
+
+For ``Protocol`` definitions, the most consistent interpretation would be that the only classes that can implement the
+protocol would be classes that use nominal inheritance from the protocol, or ``@final`` classes that implement the protocol.
+Other classes either have or could potentially have a solid base that is not the protocol. This is convoluted and not useful,
+so we disallow ``@solid_base`` on ``Protocol`` definitions.
+
+Similarly, the concept of a "solid base" is not meaningful on ``TypedDict`` definitions, as TypedDicts are purely structural types.
+
+Although they receive some special treatment in the type system, ``NamedTuple`` definitions create real nominal classes that can
+have child classes, so it makes sense to allow ``@solid_base`` on them and treat them like regular classes for the purposes
+of the solid base mechanism.
```

Comment:
I agree that there's no good reason to disallow `@solid_base` on a `NamedTuple` class. It might be worth noting, though, that any `NamedTuple` classes already has a solid base in its MRO, since all `NamedTuple` classes inherit from `tuple`, and `tuple` is a solid base.

Reply:
Added